Allow (subdir ...) in dune files used via (include ...)

nojb opened the pull request:
Currently, the `(include ...)` stanza is processed after the `(subdir ...)` stanzas have been removed from the input. This makes it impossible to support `(subdir ...)` in a file used via `(include ...)`. The PR adjusts things so that the `(include ...)` stanza is processed simultaneously with `(subdir ...)`, so that this is allowed.

I found the code that handles `(subdir ...)` a bit tricky, so would appreciate a review by someone familiar with it. Thanks!

Fixes #3419

rgrinberg commented:
Doesn't this break generating `include` in ocaml syntax dune files?

nojb commented:
> Doesn't this break generating `include` in ocaml syntax dune files?

Good catch, it does :)

Will look into it.

rgrinberg commented:
We'll probably need to evaluate `include` in both stages (file_tree and dune_file).

rgrinberg commented:
> I found the code that handles (subdir ...) a bit tricky, so would appreciate a review by someone familiar with it. Thanks!

Sorry about that. I can try to add some comments to the parts you found tricky.

nojb commented:
> > I found the code that handles (subdir ...) a bit tricky, so would appreciate a review by someone familiar with it. Thanks!
> 
> Sorry about that. I can try to add some comments to the parts you found tricky.

No need! I meant that it took me some time to figure out how the logic for `(subtree ...)` worked, but the code itself is very clear :)

nojb commented:
> We'll probably need to evaluate `include` in both stages (file_tree and dune_file).

I pushed a commit doing this.

rgrinberg commented:
There seems to be a failure in CI:

```
File "test/blackbox-tests/test-cases/dialects.t/bad1/dune-project", line 9, characters 1-74:
395 9 |  (name d)
39610 |  (implementation (extension foo2))
39711 |  (interface (extension bar2)))
398Error: dialect "d" is already defined
```

nojb commented:
> There seems to be a failure in CI:

Ya, I'm looking into it, thanks.

nojb commented:
> There seems to be a failure in CI:
> 
> ```
> File "test/blackbox-tests/test-cases/dialects.t/bad1/dune-project", line 9, characters 1-74:
> 395 9 |  (name d)
> 39610 |  (implementation (extension foo2))
> 39711 |  (interface (extension bar2)))
> 398Error: dialect "d" is already defined
> ```

It is strange: this is the expected output of the test. On the other hand, I can only reproduce the problem in this branch, not in `master`...

nojb commented:
I think I understand what is happening: with this patch, `(include ...)` is evaluated more eagerly. In particular, it seem that the include in `test/blackbox-tests/dune` is being loaded too eagerly, which causes the error above as soon as `dune` is being built.

nojb commented:
OK, I think I found the bug: I am not correctly merging the `data_only_dir` stanzas from the included files.

nojb commented:
OK, I reworked the patch to fix the problem. I modified the approach a bit: now I process `(include ...)` in a separate pass before processing `(subdir ...)` and the rest of the stanzas. The first pass removes any `(include ...)` node and replaces it by the contents of the included file in the AST, which is then passed onwards and processed as usual.

nojb commented:
I rebased and cleaned up the history. This is ready for review. Thanks!

nojb commented:
> Looks good. I'm surprised how simple this turned out to be.

Thanks for the quick review!

> This needs a CHANGES entry.

Done!

nojb commented:
I realised that there may be a small adjustment to make. Suppose that we have:
```
(subdir a (include dune.inc))
```
Is the file `dune.inc` supposed to be found inside `a` or in the current directory? At the moment, the PR looks for it within the current directory, not within `a`. It should be easy to adjust, but wanted to confirm first.

rgrinberg commented:
It should be relative to directory “a" I think

Rudi.
On Aug 5, 2020, 10:55 PM -0700, Nicolás Ojeda Bär <notifications@github.com>, wrote:
> I realised that there may be a small adjustment to make. Suppose that we have:
> (subdir a (include dune.inc))
> Is the file dune.inc supposed to be found inside a or in the current directory? At the moment, the PR looks for it within the current directory, not within a. It should be easy to adjust, but wanted to confirm first.
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or unsubscribe.

nojb commented:
> It should be relative to directory “a" I think Rudi.

Indeed, it's that. I pushed a commit (and a corresponding test) for this. It should be good to go now.

rgrinberg commented:
@nojb I just talked to Jeremie and we decided that it's best to gate this feature behind 2.7. If users end up relying on this feature without putting the appropriate constraint in their opam file, it will lead to some issues down the road.

If it's not too hard, could you add the version check?

rgrinberg commented on src/dune/sub_dirs.ml:
```diff
@@ -307,16 +307,19 @@ let decode =
-    let+ nodes = fields (decode ~context ~path) in
+    let+ nodes = fields (decode ~context ~path ~inside_include) in
+    if inside_include && dune_version < (2, 7) then
+      Dune_lang.Syntax.Error.since loc Stanza.syntax (2, 7) ~what:"Using a `subdir' stanza within an `include'd file";
```
`$ make fmt` is needed perhaps?

nojb replied:
Yes, done!

nojb commented:
Version check (+ test) added.

rgrinberg commented:
@nojb are you still working on this PR?

nojb commented:
> @nojb are you still working on this PR?

No, the PR is ready to go as far as I am concerned.

rgrinberg commented:
Okay. Then I will merge it today unless you'll do the honors and do it before me.

nojb commented:
> Okay. Then I will merge it today unless you'll do the honors and do it before me.

Thanks. I'll let you do it; I am away from home with horrible internet service :)
